### ISSUE TITLE
Displaying command specific help

### ISSUE THREAD

**Issue by mikeerickson** [infinitered/gluegun#427]
Does any interface exist to display command specific help?

**Comment by jamonholmgren** [comment]
@mikeerickson Hm, good question. Right now, no, but that's something we should add.

Something like:

```shell
$ movie imdb --help
```

... should display help for all commands under `./src/commands/imdb/`.

For now, you can check it in your root command:

```typescript
import { GluegunToolbox } from 'gluegun'

const HELP_MESSAGE = `
movie api
      
  Run \`movie api reset\` to reset your IMDB token.
      
`

module.exports = {
  name: 'api',
  run: async (toolbox: GluegunToolbox) => {
    const {
      meta: { commandInfo },
      print,
      parameters: { options }
    } = toolbox

    if (options.help || options.h) {
      print.info(HELP_MESSAGE)
      return
    }

    // do other things
  }
}
```

There is already a built-in capability to do this here:

https://github.com/infinitered/gluegun/blob/663972d05a01cb83852a39486a6f5f97abf7ad5b/src/toolbox/meta-tools.ts#L51-L66

However, it's not exposed with the full interface:

https://github.com/infinitered/gluegun/blob/663972d05a01cb83852a39486a6f5f97abf7ad5b/src/core-extensions/meta-extension.ts#L17

If we updated `toolbox.meta.commandInfo()` to take an argument and pass that on as the second argument, then it would be easy to implement this:

```
    const { meta, print } = toolbox
    print.info(meta.commandInfo(['api']))
```

We should also make it so this is essentially what happens when you do `mycli foo bar --help` by default, especially if you added the `.help()` call to the CLI builder.

I definitely want to do this.

**Comment by RichiCoder1** [comment]
Being able to print what arguments a command expects would be nice too. Should I make that a separate ticket?

**Comment by mikeerickson** [comment]
I think these two Items go hand in hand, definitely something that should be supported together. Same ticket, or separate but both should be supported.

**Comment by mikeerickson** [comment]
@RichiCoder1 @jamonholmgren 

Back from holiday, figured I would chime in some more on this topic.  

I have created CLIs internally using some homegrown libraries and one of the important features is the ability to show command specific help, including support for command attributes

This is how command specific help appears:

![image](https://user-images.githubusercontent.com/183153/50607861-967e3600-0e7f-11e9-9741-746bd72e36ff.png)

The content for command help is read from the command file

```js

module.exports = {
  init: function(cli) {
    cli.arguments.commitlint = cli.arguments.c = cli.arguments.commitlint || cli.arguments.c || true;
  },
  name: "husky",
  disabled: false,
  description: "Install and configure husky module",
  flags: {
    "--commitlint, -c": "configures commitlint"
  },
  run: function(cli) {
    cli.arguments = cli.setDefaultFlags(cli, this.flags);

    this.hasOwnProperty("init") ? this.init(cli) : null;

    cli.print.info(`⚙️  Execute ${this.name} command`);
    cli.print.info(cli.strings.stringify(cli.arguments));
  }
};

```

**Comment by mikeerickson** [comment]
And for completeness sake, here is what the default help looks like (I am thinking of creating a plugin for gluegun to show this as default as opposed to current help)  Or, I could just create a simple function for help interface to use

![image](https://user-images.githubusercontent.com/183153/50608018-13111480-0e80-11e9-8069-faa8005fcf51.png)

**Comment by jamonholmgren** [comment]
I really love that @mikeerickson. Any help you can provide is greatly appreciated, as I'm focusing on some other aspects of Gluegun at the moment.

**Comment by mikeerickson** [comment]
I will surely add the necessary code to provide this functionality.  Would you like me to implement similar UI to what I have shown above.

I was going to create a new plugin for personal use, but if you are keen to having this UI in core, I would prefer to do it there as well.

**Comment by mikeerickson** [comment]
@jamonholmgren sounds good, I have already started the "Getting to know the source" process. I will do the primary help interface, then move onto the command help.

I am reviewing the cli.integration.ts tests right now.

**Comment by jamonholmgren** [comment]
Feel free to join the #gluegun channel in http://community.infinite.red Slack to chat about this in realtime, Mike.

**Comment by mikeerickson** [comment]
@jamonholmgren I have already joined :-)

**Comment by mikeerickson** [comment]
@jamonholmgren OK, first pass is done (prettify the main cli help)   I have not fixed the tests yet, but wanted to share this with you. There will be a few more tweaks (ie adding the Usage section, not sure how that is going to be implemented just yet)

![image](https://user-images.githubusercontent.com/183153/50621791-1f679280-0ebd-11e9-8865-839149fab6bb.png)

**Comment by RichiCoder1** [comment]
Looks awesome! Bit late, but let me know if yah need any help.

**Comment by mikeerickson** [comment]
@RichiCoder1 So far, so good.  Getting up to speed with the source code is the longest part, but that is going swimmingly.

One question, do you have place where you store application wide constants?  I am sure I could track it down, but since you are here :-)

**Comment by RichiCoder1** [comment]
I'm as a novice to the code base as you :). But from what I've seen so far, constants have been kept close to the most relevant place they're used. So `colors` in `print-tools`, plugin defaults with `plugin-loader`, etc. Current @jamonholmgren?

**Comment by mikeerickson** [comment]
@jamonholmgren Here are my proposed changes to the command object.  I have only added `flags` property to the command object (if it is not supplied, it will operate as normal).   Let me know if there are any other properites you might need on the `flags` property

```
module.exports = {
  name: 'prompty',
  description: 'Get user information',
  flags: {
    age: {
      alias: ['a', 'age'],
      description: 'Supply age',
    },
    shoe: {
      alias: ['s', 'shoe'],
      description: 'What shoes are you wearing',
      choices: ['Clown', 'Adidas', 'Nike'],
      initial: 'Adidas',
    },
  },
  run: async function(toolbox) {
    const { print } = toolbox

    print.info('Welcome to your CLI')
    let choices = this.flags.shoe.choices || ['Other']

    // let choices = this.flags.shoe.choices
    // text input
    const askAge = { type: 'input', name: 'age', message: 'How old are you?' }

    // multiple choice
    const askShoe = {
      type: 'list',
      name: 'shoe',
      message: 'What shoes are you wearing?',
      choices,
    }
    // ask a series of questions
    const questions = [askAge, askShoe]
    const { age, shoe } = await toolbox.prompt.ask(questions)
    console.log(age, shoe)
  },
}
```

**Comment by jamonholmgren** [comment]
It looks awesome.

Minor nit: would you need to put `shoe` in the `alias` list if it's already the name of the flag? Perhaps just any aliases that aren't the name of the flag.

Also, you may want `dashed: true|false` as an option, since some flags might be `--shoe=Adidas` or `-s Adidas` (accessible in `toolbox.parameters.options['shoe'] || toolbox.parameters.options['s']`).

**Comment by mikeerickson** [comment]
@jamonholmgren Are the CLI arguments available on the `toolbox` object anywhere?  Just thinking ahead as to how the user will access these arguments, or will I need to add code for that as well.

**Comment by jamonholmgren** [comment]
@mikeerickson Apologies, I've been out of town. I'm not sure what you're asking, to be honest -- there is [toolbox.parameters](https://github.com/infinitered/gluegun/blob/master/docs/toolbox-parameters.md), is that what you're looking for?

**Comment by danawoodman** [comment]
Is this dead? Is there any way to auto-document `toolbox.parameters.options`?

**Comment by jamonholmgren** [comment]
@danawoodman Currently I don't have plans to add it (we haven't needed it for our CLIs yet), but if you have any ideas, I'm happy to review!

**Comment by danawoodman** [comment]
@jamonholmgren I don't mind the proposals discussed here. Some way of passing in a definition of the options/flags that the given command supports then those are automatically documented/validated before hand. I find I'm writing a LOT of extra code to do this instead of Gluegun doing which I would hope.

Thinking out-loud:
```
options: [
  { option: '-y, --yes', type: boolean, default: true },
  { option: '-c, --config', type: string, required: true },
]
```

**Comment by jamonholmgren** [comment]
I like this. We could possibly use `yargs` for this. It's currently only a devDependency, though -- we're using `yargs-parser` in production which has [some of this capability](https://github.com/yargs/yargs-parser#requireyargs-parserargs-opts) but not all.

yargs has a pretty full featured capability:

https://github.com/yargs/yargs/blob/HEAD/docs/api.md

Perhaps see if we want to bring in all of `yargs` (if it's worth the extra kbs)?

**Comment by danawoodman** [comment]
@jamonholmgren yeah, to me a few kb is fine esp since my CLI has to bundle the whole of Node so it already is bloated haha

Maybe the support is optional? Might be confusing though.

A way to be able to define typed options and have the CLI output help for them would be killer

**Comment by HurricaneInteractive** [comment]
I managed to get a "auto-generated" help command to work without modifying Gluegun or adding `yargs`. I will say though, it is quite basic and doesn't support typed options. Another thing is, you have to run a function at the start of each command in order to capture the `--help, -h` flag.

I have created a [gist](https://gist.github.com/HurricaneInteractive/a4808bf2f2fca507271a5a465899e7ae) (using TypeScript) with how I achieved it. Feel free to use it as a starting point. Again, very basic implementation and only displays the command, your description and available flags. If you use the files in the gist, when you run `<brand> some command --help`, you should get.

![Screen Shot 2020-07-09 at 2 09 17 pm](https://user-images.githubusercontent.com/25698413/86996070-00526b00-c1ee-11ea-889a-83f2ecb77976.png)

**Comment by benhickson** [comment]
+1 interested to see how this develops! :)

**Comment by jamonholmgren** [comment]
I realize this is an old issue, and not much progress has happened, but I do intend to work on it eventually. Perhaps a good option for hacking [on my Twitch stream](https://twitch.tv/jamonholmgren).
